manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,993 @@
+		}
+	// case api.EventUpdateNode:
+	//     we don't need the update node case because nothing about
+	//     a node update will require reallocation.
+	case api.EventDeleteNode:
+		if ev.Node != nil {
+			if _, ok := a.deletedObjects[ev.Node.ID]; ok {
+				delete(a.deletedObjects, ev.Node.ID)
+				return
+			}
+			a.network.DeallocateNode(ev.Node)
+			delete(a.pendingNodes, ev.Node.ID)
+		}
+	}
+}
+
+func (a *NewAllocator) processPendingAllocations(ctx context.Context) {
+	ctx = log.WithField(ctx, "method", "(*NewAllocator).processPendingAllocations")
+
+	// we need to hold the lock the whole time we're allocating.
+	a.pendingMu.Lock()
+	defer a.pendingMu.Unlock()
+
+	log.G(ctx).Debugf(
+		"processing pending allocations. %v networks, %v nodes, %v tasks",
+		len(a.pendingNetworks), len(a.pendingNodes), len(a.pendingTasks),
+	)
+
+	defer func() {
+		log.G(ctx).Debugf(
+			"finished processing pending allocations. retrying: %v networks, %v nodes, %v tasks",
+			len(a.pendingNetworks), len(a.pendingNodes), len(a.pendingTasks),
+		)
+	}()
+
+	// collect metrics on how long each batch of pending allocations takes.
+	// start at timer after we acquire the lock, and then defer the done call.
+	done := metrics.StartTimer(batchProcessingDuration)
+	defer done()
+
+	// overlayAllocated is a variable that tracks whether a new overlay network
+	// has been allocated. If it is true, then regardless of which nodes are
+	// currently pending, all nodes will need to be reallocated with the new
+	// network. we will do this after network allocation, but before task
+	// allocation, to prefer the case where nodes exhaust all available IP
+	// addresses leaving some tasks without. Tasks can just be failed, but
+	// nodes need IP addresses
+	overlayAllocated := false
+
+	// capture the pending maps and reinitialize them, so they'll be clean
+	// after this runs
+	pendingNetworks := a.pendingNetworks
+	a.pendingNetworks = map[string]struct{}{}
+	pendingNodes := a.pendingNodes
+	a.pendingNodes = map[string]struct{}{}
+	pendingTasks := a.pendingTasks
+	a.pendingTasks = map[string]struct{}{}
+	// first, pending networks. open the store and grab all of the pending
+	// networks. networks can't be updated, so we don't need to worry about a
+	// race condition.
+
+	// if no networks are pending, skip this block. this saves us the time it
+	// takes to acquire the store lock
+	if len(pendingNetworks) > 0 {
+		// we'll need two slices. the first is to hold all of the networks we get
+		// from the store. the second holds all of the successfully allocated
+		// networks
+		networks := make([]*api.Network, 0, len(pendingNetworks))
+		allocatedNetworks := make([]*api.Network, 0, len(pendingNetworks))
+
+		batchDone := metrics.StartTimer(storeLockHeld.WithValues("networks", "allocate", "read"))
+		a.store.View(func(tx store.ReadTx) {
+			for id := range pendingNetworks {
+				// remove the network from the pending allocations. if something
```

Comment:
This comment no longer seems applicable, at least not in this particular location.  We just start off removing all networks from `a.pendingNetworks`, and when they are re-added you already have comments in those locations stating why.

Reply:
Removed.